### ISSUE TITLE
jobs frequently stopping with "session not active"

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#498]
In GitLab by @ rbernard on Mar 2, 2020, 10:53

Hi,
jobs are frequently stopping (2-3 times a day) with the same message, see:

/esarchive/autosubmit/a2er/outa2er.txt.example

using default autosubmit module.

thanks

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 2, 2020, 11:16

Hi @ rbernard ,

I recommend you to use at least version 3.12.0b since it will be the default version soon and only is updated for bug fixes ( and it is the default for auto-monarch and auto-ecearth.

The issue itself, the issue is that the login node specified on the platform.conf seems to have some crashes I recommend to use other login nodes until the issues are solved from the infrastructure side.

In addition, In version 3.12.1b which is weekly updated so is not as stable as 3.12.0b, you can specify multiple login nodes https://earth.bsc.es/gitlab/es/autosubmit/issues/411 and will try always to use the first node and use the others as a fallback.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 5, 2020, 11:18

closed

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 7, 2020, 09:20

reopened

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 7, 2020, 09:22

this keeps happening regardless of the login node. See 
/esarchive/autosubmit/a2mc/outa2mc.txt.example2

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 7, 2020, 14:02

I did a search in all the AS logs for this year and the error is not happening so often, so I would discard that indeed is caused by real connection issues (if that was the case, we should see the error in different experiments running at the same time).

Could it be a Paramiko bug @dbeltrankyl ? Maybe caused by the timeout?

https://stackoverflow.com/questions/36490989/how-to-keep-ssh-session-not-expired-using-paramiko

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 7, 2020, 15:59

I will take a look at it during this week, @ rbernard it happens when AS is running during a large amount of time?

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 7, 2020, 16:12

it happens every one/two days

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 8, 2020, 09:55

Any other experiments? I checked a2mc and I only see messages the 7th july.

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 8, 2020, 10:04

it happened all the time with the omip experiments that are now
concluded ( a2er, a2l5, a2mc, a2md)

I am using a2mc to extend one of the omip exp. Currently it has finished
one segment but I'll re-launch it this morning

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 8, 2020, 10:11

Thanks @ rbernard , I was looking yesterday and I didn't find the same message that you reported for a2mc in many experiments.

But I found another type of message in a2l5:

```
[ERROR] 2020-03-21 00:19:16,567 Can not send command mkdir -p /gpfs/scratch/pr1eki00/pr1eki01/a2l5/LOG_a2l5 to mn2-pr1eki00: SSH session not active
[ERROR] 2020-03-21 00:19:16,567 Could not create the DIR /gpfs/scratch/pr1eki00/pr1eki01/a2l5/LOG_a2l5 on HPC mn2-pr1eki00
2020-03-21 00:19:16,567 Could not remove file /gpfs/scratch/pr1eki00/pr1eki01/a2l5/LOG_a2l5/a2l5_19480101_fc00_57_CMOROCE.cmd
[ERROR] 2020-03-21 00:19:16,568 Can not send file /esarchive/autosubmit/a2l5/tmp/a2l5_19480101_fc00_57_CMOROCE.cmd to /gpfs/scratch/pr1eki00/pr1eki01/a2l5/LOG_a2l5/a2l5_19480101_fc00_57_CMOROCE.cmd
[ERROR] 2020-03-21 00:19:16,568 marenostrum4 submission failed
[CRITICAL] 2020-03-21 00:19:16,895 Unhandled exception on Autosubmit: SSH session not active
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/autosubmit.py", line 375, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/autosubmit.py", line 790, in run_experiment
    if Autosubmit.submit_ready_jobs(as_conf, job_list, platforms_to_test, packages_persistence):
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/autosubmit.py", line 833, in submit_ready_jobs
    package.submit(as_conf, job_list.parameters)
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/job/job_packages.py", line 87, in submit
    self._send_files()
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/job/job_packages.py", line 115, in _send_files
    self.platform.send_file(self._job_scripts[job.name])
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/platforms/paramiko_platform.py", line 128, in send_file
    ftp = self._ssh.open_sftp()
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/client.py", line 379, in open_sftp
    return self._transport.open_sftp_client()
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/transport.py", line 811, in open_sftp_client
    return SFTPClient.from_transport(self)
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/sftp_client.py", line 127, in from_transport
    max_packet_size=max_packet_size)
  File "/shared/earth/software/autosubmit/3.11.0-hotfix-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/transport.py", line 615, in open_session
    max_packet_size=max_packet_size)
SSHException: SSH session not active
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 8, 2020, 10:30

I understand, I was looking for errors in July, this is from March.

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 8, 2020, 10:43

yes, that's when I opened this issue which was then closed without a solution ( I kept having the same problem despite trying what @dbeltrankyl suggested). I finished those experiments and didn't see the problem until now when I started re-running one of those experiments. So, I re-opened the issue.

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 8, 2020, 10:49

I've made a bit of confusion here. Sorry.

the original problem is still happening, although right now I don't have any fresh "proof" of it because I have overwritten all logs, the error message is the one here (from an old log):

/esarchive/autosubmit/a2mc/outa2mc.txt.example

the error message here:

/esarchive/autosubmit/a2mc/outa2mc.txt.example2

I have seen it only in that instance - i.e. it's not recurrent

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 8, 2020, 11:34

Thanks for the details @ rbernard !

We will look if it's a timeout issue with the Autosubmit connector, or if otherwise is an issue caused by a real connection issue.

In the first case maybe a bigger timeout or another workaround could help. In the second case, in 3.13.0 we support adding multiple logins for the same platform, so in this situation Autosubmit should try using another login (I know you are not using this version, but the idea is to prevent it happens when we move to the new version in few months).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 8, 2020, 16:31

Hi,

There is no timeout set in connection, if not set it should be unlimited(according to the docs) it is only set for a particular command and doesn't seem related to the "session not active" if that error happen more that twice as shown in the logs, I investigated a bit on the paramiko git and some people have this same issue however it was not resolved or was with a different scenario. 

Also, all these session active errors that I found were with older versions of paramiko so perhaps it has been solved on the latest version (we are behind 1 mayor release (1.15 vs 2.7.1) and soon 2 major release maybe I can try to update it? I did for the alpha of autosubmit 4 and didn't seem that the build had issues.

The retry mechanism should solve this on 3.13.0b, maybe I should launch a long test to see if it still happens on 3.13.0b @mcastril ?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 8, 2020, 16:47

Thanks @dbeltrankyl 

> The retry mechanism should solve this on 3.13.0b, maybe I should launch a long test to see if it still happens on 3.13.0b @mcastril ? 

Yes, you maybe can try to reproduce the error by launching an experiment and try to force the connection problem, to see if the error is the same. 

There are soft to simulate slow connections like https://wiki.linuxfoundation.org/networking/netem . Another option is maybe to try using a VM.

**Comment by kinow** [comment]
In GitLab by @ rbernard on Jul 9, 2020, 16:51

hi, it just happened again:

/esarchive/autosubmit/a2mc/outa2mc.txt.example3

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 13, 2020, 09:41

I had this issue on 3.13.0b also, it should try to reconnect into a different login but keep stuck on retrials while even if the retrial value is negative.

About the code, the connection socket has an infinite timeout but still, it is being lost somehow.

So, for 3.13.0b the plan is to bugfix reconnect work when this happens (to the same login or a different one)

**Comment by kinow** [comment]
In GitLab by @ abunuel on Aug 31, 2020, 17:44

Hi, i had a similar error with my jobs (a32v, a31h). They stopped with a message: Can not send command sacct -n -X
      >      -j 11493592 -o "State" to prace-mn: exceptions must be old-style classes
      >      or derived from BaseException, not NoneType
Now i've re-run the exps to see if this message appears again.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 27, 2020, 13:27

Closing this issue because 3.12.0 was patched and 3.13.0 should not suffer from this problem:

> Related to #601 ,#600 https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1427#note_100744
>
> I created a patch on a new module called `autosubmit/3.12.0-fixed-foss-2015a-Python-2.7.9`,
>
> Please, the users that had issues with 3.12.0 current version, swap to this version to see if the changes are correct and do not break anything if it goes well it will be the default one.
>
> The module is essentially the same than 3.12.0 but instead of hanging all the experiment what will now occur now in these cases that job, in particular, will be set into UNKNOWN status. if you have the mail notifier set on, it will send you an email whenever this issue happens.
>
> If autosubmit can detect later the real status of the job it will try to put the correct status, if not you will need to do it manually.
>
> Apologies for the issues caused, this is just a non-invasive patch, further improvements will be on 3.13.0b

https://earth.bsc.es/gitlab/es/autosubmit/-/issues/601

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 27, 2020, 13:27

closed
